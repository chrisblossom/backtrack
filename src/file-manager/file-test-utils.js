/* @flow */

/* eslint-disable arrow-body-style */

import type { FileManager } from '../types.js';

const filesPreprocessor = (args) =>
    // $FlowIgnore
    require('../options-file/files-preprocessor').filesPreprocessor(args);
const filesParser = (args) =>
    // $FlowIgnore
    require('../options-file/files-post-processor').filesPostProcessor(args);
const getFileStats = (parsedFiles) =>
    require('./get-file-stats').getFileStats(parsedFiles);

function fileStats(previousFiles: FileManager) {
    const normalize = filesPreprocessor({ value: previousFiles });

    const parsedPreviousFiles = filesParser({ value: normalize });

    const previousStats = getFileStats(parsedPreviousFiles).files;

    return previousStats;
}

function fileInfo(files: FileManager = [], previousFiles: FileManager = []) {
    const previousStats = fileStats(previousFiles);
    const normalize = filesPreprocessor({ value: files });

    const parsedFiles = filesParser({ value: normalize });

    return {
        previousStats,
        parsedFiles,
    };
}

export { fileInfo, fileStats };
