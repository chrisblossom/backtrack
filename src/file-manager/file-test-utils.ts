/* eslint-disable arrow-body-style,@typescript-eslint/no-unsafe-member-access,@typescript-eslint/no-unsafe-return,@typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-assignment */

import { FileManager, FileManagerStats } from '../types';

const filesPreprocessor = (args: any) =>
	require('../options-file/files-preprocessor').filesPreprocessor(args);
const filesParser = (args: any) =>
	require('../options-file/files-post-processor').filesPostProcessor(args);
const getFileStats = (parsedFiles: any) =>
	require('./get-file-stats').getFileStats(parsedFiles);

function fileStats(previousFiles: FileManager): FileManagerStats {
	const normalize = filesPreprocessor({ value: previousFiles });

	const parsedPreviousFiles = filesParser({ value: normalize });

	const previousStats = getFileStats(parsedPreviousFiles).files;

	return previousStats;
}

interface FileInfo {
	previousStats: FileManagerStats;
	parsedFiles: FileManager;
}

function fileInfo(
	files: FileManager = [],
	previousFiles: FileManager = [],
): FileInfo {
	const previousStats = fileStats(previousFiles);
	const normalize = filesPreprocessor({ value: files });

	const parsedFiles = filesParser({ value: normalize });

	return {
		previousStats,
		parsedFiles,
	};
}

export { fileInfo, fileStats };
