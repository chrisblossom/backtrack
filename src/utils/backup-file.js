/* @flow */

import path from 'path';
import { existsSync } from 'fs';
import { move } from 'fs-extra';
import { getFileHash } from './get-file-hash';

import { rootPath } from '../config/paths';

type Return = Promise<?{
    file: string,
    filename: string,
}>;

async function backupFile(file: string): Return {
    const hash = getFileHash(file);

    const { base: filename, dir, ext = '' } = path.parse(file);

    const shortHash = hash.slice(0, 8);

    let backupFilename = `${filename}-backup-${shortHash}${ext}`;
    let backupFilePath = path.resolve(dir, backupFilename);

    /**
     * Handle the same file being backed up multiple times
     */
    let previouslyBackedUp = false;
    let backupFileExists = existsSync(backupFilePath);
    if (backupFileExists) {
        let count = 0;
        const maxCount = 10;
        while (
            backupFileExists &&
            previouslyBackedUp === false &&
            count < maxCount
        ) {
            const backupFileHash = getFileHash(backupFilePath);

            /**
             * Check if the existing file has the same hash.
             *
             * Prevents backing up the same file over
             */
            if (backupFileHash === hash) {
                previouslyBackedUp = true;
            }

            backupFilename = `${filename}-backup-${shortHash}-${count}${ext}`;
            backupFilePath = path.resolve(dir, backupFilename);

            backupFileExists = existsSync(backupFilePath);
            count += 1;
        }

        /**
         * Prevent flooding filesystem with backups
         */
        if (count === maxCount && backupFileExists) {
            const relativeFile = path.relative(rootPath, file);

            throw new Error(`Remove all existing backups of: ${relativeFile}`);
        }
    }

    /**
     * Don't backup anything if the file has already been backed up
     */
    if (previouslyBackedUp === false) {
        /**
         * Will fail if backup file already exists
         */
        await move(file, backupFilePath, {
            overwrite: false,
        });

        return {
            file: backupFilePath,
            filename: backupFilename,
        };
    }

    return undefined;
}

export { backupFile };
