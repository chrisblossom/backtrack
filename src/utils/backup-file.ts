import path from 'path';
import { existsSync } from 'fs';
import { move } from 'fs-extra';
import { rootPath } from '../config/paths';
import { getFileHash } from './get-file-hash';

type Return = Promise<
	| {
			file: string;
			filename: string;
	  }
	| undefined
>;

async function backupFile(file: string): Return {
	const hash = getFileHash(file);

	const { base: filename, dir, ext = '' } = path.parse(file);

	const shortHash = hash.slice(0, 8);

	let backupFilename = `${filename}-backup-${shortHash}${ext}`;
	let backupFilePath = path.resolve(dir, backupFilename);

	/**
	 * Handle the same file being backed up multiple times
	 */
	let previouslyBackedUp = false;
	let backupFileExists = existsSync(backupFilePath);
	if (backupFileExists) {
		let count = 0;
		const maxCount = 10;
		while (
			backupFileExists &&
			previouslyBackedUp === false &&
			count < maxCount
		) {
			const backupFileHash = getFileHash(backupFilePath);

			/**
			 * Check if the existing file has the same hash.
			 *
			 * Prevents backing up the same file over
			 */
			if (backupFileHash === hash) {
				previouslyBackedUp = true;
			}

			backupFilename = `${filename}-backup-${shortHash}-${count.toString()}${ext}`;
			backupFilePath = path.resolve(dir, backupFilename);

			backupFileExists = existsSync(backupFilePath);
			count += 1;
		}

		/**
		 * Prevent flooding filesystem with backups
		 */
		if (count === maxCount && backupFileExists) {
			const relativeFile = path.relative(rootPath, file);

			throw new Error(`Remove all existing backups of: ${relativeFile}`);
		}
	}

	/**
	 * Don't backup anything if the file has already been backed up
	 */
	if (previouslyBackedUp === false) {
		/**
		 * Will fail if backup file already exists
		 */
		await move(file, backupFilePath, {
			overwrite: false,
		});

		return {
			file: backupFilePath,
			filename: backupFilename,
		};
	}

	return undefined;
}

export { backupFile };
