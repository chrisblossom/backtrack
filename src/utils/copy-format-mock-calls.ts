/**
 * copy.mock.calls is async so order is not guaranteed for expected results
 * Converts to plain object
 *
 * For testing purposes only
 *
 * meant to be used as a snapshot with jest-serializer-path
 */

type CopyCalls = [string, string, {}];

function copyFormatMockCalls(calls: ReadonlyArray<CopyCalls>) {
	const result = calls.reduce((acc, call) => {
		const src = call[0];
		const dest = call[1];
		const options = call[2];

		return {
			...acc,
			[src]: {
				dest,
				options,
			},
		};
	}, {});

	return result;
}

export { copyFormatMockCalls };
