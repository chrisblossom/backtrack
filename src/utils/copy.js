/* @flow */

import fse from 'fs-extra';
import { toArray } from './object-utils';

import type { CopyFile } from '../types.js';

/**
 * See https://github.com/jprichardson/node-fs-extra/blob/master/docs/copy.md#copysrc-dest-options-callback
 * for options
 */
async function copy(files: $ReadOnlyArray<CopyFile> | CopyFile, options: {}) {
    const normalized = toArray(files);

    /**
     * Does not return any information
     */
    await Promise.all(
        normalized.map((file) => {
            return fse.copy(file.src, file.dest, options);
        }),
    );
}

export { copy };
