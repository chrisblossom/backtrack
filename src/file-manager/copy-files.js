/* @flow */

import path from 'path';
import { copy } from '../utils/copy';
import log from '../utils/log';
import { rootPath } from '../config/paths';

import type { ParsedFiles, FileStats } from '../types.js';

async function copyFiles(files: ParsedFiles, previousStats?: FileStats = {}) {
    const src = files.src;
    const dest = files.dest;

    const filesToCopy = src.files.reduce((acc, srcFile, index) => {
        const destFile = dest.files[index];

        const srcHash = src.hash[srcFile];
        const destHash = dest.hash[destFile];

        /**
         * Only copy files that have changed / do not exist
         *
         * hash === '' means file does not exist
         */
        if (srcHash !== destHash) {
            const srcAbsolute = src.absolute[srcFile];

            if (destHash) {
                const previousHash = previousStats[destFile];
                /**
                 * Do not copy if changes are allowed and destFile has been modified
                 */
                const allowChanges = dest.allowChanges[destFile];
                if (allowChanges === true && destHash !== previousHash) {
                    const destAbsolute = dest.absolute[destFile];
                    const { base: filename, dir, ext = '' } = path.parse(
                        destAbsolute,
                    );

                    const updatedFilename = `${filename}-latest${ext}`;

                    const updatedFileAbsolute = path.resolve(
                        dir,
                        updatedFilename,
                    );

                    const relativeFilePath = path.relative(
                        rootPath,
                        updatedFileAbsolute,
                    );

                    /**
                     * Since backtrack-stats tracks the latest srcHash,
                     * only copy latest file if source has changed.
                     */
                    if (srcHash === previousHash) {
                        return acc;
                    }

                    log.info(
                        `Changed file ${destFile} source updated. Copying to: ${relativeFilePath}`,
                    );

                    const copyFile = {
                        src: srcAbsolute,
                        dest: updatedFileAbsolute,
                    };

                    return [...acc, copyFile];
                }

                log.info(`Updating file: ${destFile}`);
            } else {
                log.info(`Adding file: ${destFile}`);
            }

            const destAbsolute = dest.absolute[destFile];
            const copyFile = {
                src: srcAbsolute,
                dest: destAbsolute,
            };

            return [...acc, copyFile];
        }

        return acc;
    }, []);

    /**
     * Bail early if nothing else to do
     */
    if (filesToCopy.length === 0) {
        return;
    }

    /**
     * Copy all files
     *
     * Must overwrite because source files can change
     * Safe to overwrite because changed dest files have been backed up
     */
    await copy(filesToCopy, {
        overwrite: true,
        errorOnExist: true,
    });
}

export { copyFiles };
