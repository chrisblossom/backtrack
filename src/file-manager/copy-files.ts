import path from 'path';
import { copy, File } from '../utils/copy';
import log from '../utils/log';
import { rootPath } from '../config/paths';
import { ParsedFiles, FileStats } from '../types';

// prettier-ignore
type CopyInfo =
	{ shouldCopy: false } |
	{ shouldCopy: true; destPath: string };

/*
 * Handles the logic for when file changes are allowed.
 * It determines if a file should be copied and creates a new filename with '-latest'.
 */
function handleAllowedChanges(
	files: ParsedFiles,
	destFile: string,
	srcHash: string,
	previousHash: string,
): CopyInfo {
	const destAbsolute = files.dest.absolute[destFile];
	const { base: filename, dir, ext = '' } = path.parse(destAbsolute);
	const updatedFilename = `${filename}-latest${ext}`;
	const updatedFileAbsolute = path.resolve(dir, updatedFilename);
	const relativeFilePath = path.relative(rootPath, updatedFileAbsolute);

	/**
	 * Since backtrack-stats tracks the latest srcHash,
	 * only copy latest file if source has changed.
	 */
	if (srcHash === previousHash) {
		return { shouldCopy: false };
	}

	log.info(
		`Changed file ${destFile} source updated. Copying to: ${relativeFilePath}`,
	);

	return { shouldCopy: true, destPath: updatedFileAbsolute };
}

/*
 * Determines whether a file should be copied based on its hash and other criteria.
 */
function shouldCopyFile(
	files: ParsedFiles,
	destFile: string,
	srcHash: string,
	destHash: string,
	previousStats: FileStats,
): CopyInfo {
	if (destHash) {
		const previousHash = previousStats[destFile];
		const ignoreUpdates = files.dest.ignoreUpdates[destFile];

		/**
		 * Do not copy if ignoreUpdates is true and file already exists
		 */
		if (ignoreUpdates && previousHash !== srcHash) {
			log.info(
				`Unmanaged file ${destFile} source changed. Updating .backtrack-stats.json`,
			);
			return { shouldCopy: false };
		}

		/**
		 * Do not copy if changes are allowed and destFile has been modified
		 */
		const allowChanges = files.dest.allowChanges[destFile];
		if (allowChanges && destHash !== previousHash) {
			return handleAllowedChanges(files, destFile, srcHash, previousHash);
		}

		log.info(`Updating file: ${destFile}`);
	} else {
		log.info(`Adding file: ${destFile}`);
	}

	return { shouldCopy: true, destPath: files.dest.absolute[destFile] };
}

/*
 * Creates a list of files to be copied by evaluating each source file.
 */
function getFilesToCopy(files: ParsedFiles, previousStats: FileStats): File[] {
	return files.src.files.reduce<File[]>((acc, srcFile, index) => {
		const destFile = files.dest.files[index];
		const srcHash = files.src.hash[srcFile];
		const destHash = files.dest.hash[destFile];

		if (srcHash !== destHash) {
			const srcAbsolute = files.src.absolute[srcFile];
			const destAbsolute = files.dest.absolute[destFile];

			const copyInfo = shouldCopyFile(
				files,
				destFile,
				srcHash,
				destHash,
				previousStats,
			);

			if (copyInfo.shouldCopy) {
				acc.push({
					src: srcAbsolute,
					dest: copyInfo.destPath ?? destAbsolute,
				});
			}
		}

		return acc;
	}, []);
}

/*
 * Main function to copy files.
 * It determines which files need to be copied and performs the copy operation.
 */
async function copyFiles(
	files: ParsedFiles,
	previousStats: FileStats = {},
): Promise<void> {
	const filesToCopy = getFilesToCopy(files, previousStats);

	// nothing to copy
	if (filesToCopy.length === 0) {
		return;
	}

	/**
	 * Copy all files
	 *
	 * Must overwrite because source files can change
	 * Safe to overwrite because changed dest files have been backed up
	 */
	await copy(filesToCopy);
}

export { copyFiles };
