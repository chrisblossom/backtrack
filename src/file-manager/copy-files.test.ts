import path from 'path';
import { fileInfo } from './file-test-utils';
import { copyFormatMockCalls } from '../utils/copy-format-mock-calls';

const copyFiles = (files: any, previousStats: any) =>
    require('./copy-files').copyFiles(files, previousStats);

describe('copyFiles', () => {
    const cwd = process.cwd();
    let copy: any;

    beforeEach(() => {
        copy = require.requireMock('fs-extra').copy;

        jest.mock('fs-extra', () => ({
            copy: jest.fn(() => Promise.resolve()),
        }));

        jest.mock('../utils/log', () => ({
            warn: jest.fn(),
            info: jest.fn(),
            error: jest.fn(),
            success: jest.fn(),
        }));
    });

    afterEach(() => {
        process.chdir(cwd);
    });

    test('copies only changed files', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const files = [
            {
                src: path.resolve(dir, 'nested/inside.js'),
                dest: 'nested/inside.js',
            },
            {
                src: path.resolve(dir, 'file1.js'),
                dest: 'file11.js',
            },
        ];

        const { parsedFiles, previousStats } = fileInfo(files);

        await copyFiles(parsedFiles, previousStats);

        expect(copyFormatMockCalls(copy.mock.calls)).toMatchSnapshot();
    });

    test('copies nothing if nothing changed', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const files = [
            {
                src: path.resolve(dir, 'nested/inside.js'),
                dest: 'nested/inside.js',
            },
            {
                src: path.resolve(dir, 'file1.js'),
                dest: 'file1.js',
            },
        ];

        const { parsedFiles, previousStats } = fileInfo(files);

        await copyFiles(parsedFiles, previousStats);

        expect(copy.mock.calls).toEqual([]);
    });

    test('copies all if all changed', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const files = [
            {
                src: path.resolve(dir, 'nested/inside.js'),
                dest: 'nested/inside1.js',
            },
            {
                src: path.resolve(dir, 'file1.js'),
                dest: 'file11.js',
            },
        ];

        const { parsedFiles, previousStats } = fileInfo(files);

        await copyFiles(parsedFiles, previousStats);

        expect(copyFormatMockCalls(copy.mock.calls)).toMatchSnapshot();
    });

    test('allowChanges - create file with -latest tag', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const files = [
            {
                src: path.resolve(dir, 'nested/other.js'),
                dest: 'nested/inside.js',
                allowChanges: true,
            },
            {
                src: path.resolve(dir, 'file1.js'),
                dest: 'file11.js',
            },
        ];

        const { parsedFiles, previousStats } = fileInfo(files);

        await copyFiles(parsedFiles, previousStats);

        expect(copyFormatMockCalls(copy.mock.calls)).toMatchSnapshot();
    });

    test('allowChanges - if srcHash is equal to previousHash, do nothing', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const previousFiles = [
            {
                src: path.resolve(dir, 'nested/other.js'),
                dest: 'nested/file1.js',
                allowChanges: true,
            },
        ];

        const files = [
            {
                src: path.resolve(dir, 'nested/other.js'),
                dest: 'nested/file1.js',
                allowChanges: true,
            },
        ];

        const { parsedFiles, previousStats } = fileInfo(files, previousFiles);

        await copyFiles(parsedFiles, previousStats);

        expect(copy.mock.calls).toEqual([]);
    });

    test('allowChanges - copies if does not exist', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const files = [
            {
                src: path.resolve(dir, 'nested/other.js'),
                dest: 'nested/missing.js',
                allowChanges: true,
            },
        ];

        const { parsedFiles, previousStats } = fileInfo(files);

        await copyFiles(parsedFiles, previousStats);

        expect(copyFormatMockCalls(copy.mock.calls)).toMatchSnapshot();
    });

    test('allowChanges - copies if destFile has not been modified', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const previousFiles = [
            {
                src: path.resolve(dir, 'nested/inside.js'),
                dest: 'nested/inside.js',
                allowChanges: true,
            },
        ];

        const files = [
            {
                src: path.resolve(dir, 'nested/other.js'),
                dest: 'nested/inside.js',
                allowChanges: true,
            },
        ];

        const { parsedFiles, previousStats } = fileInfo(files, previousFiles);

        await copyFiles(parsedFiles, previousStats);

        expect(copyFormatMockCalls(copy.mock.calls)).toMatchSnapshot();
    });

    test('ignoreUpdates - copies if no previous stats for file', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const files = [
            {
                src: path.resolve(dir, 'nested/other.js'),
                dest: 'nested/missing.js',
                ignoreUpdates: true,
            },
        ];

        const { parsedFiles, previousStats } = fileInfo(files);

        await copyFiles(parsedFiles, previousStats);

        expect(copyFormatMockCalls(copy.mock.calls)).toMatchSnapshot();
    });

    test('ignoreUpdates - does not copy if has previous stats', async () => {
        const dir = path.resolve(__dirname, '__sandbox__/stats1/');
        process.chdir(dir);

        const previousFiles = [
            {
                src: path.resolve(dir, 'nested/inside.js'),
                dest: 'nested/inside.js',
                ignoreUpdates: true,
            },
        ];

        const files = [
            {
                src: path.resolve(dir, 'nested/other.js'),
                dest: 'nested/inside.js',
                ignoreUpdates: true,
            },
        ];

        const { parsedFiles, previousStats } = fileInfo(files, previousFiles);

        await copyFiles(parsedFiles, previousStats);

        expect(copyFormatMockCalls(copy.mock.calls)).toMatchSnapshot();
    });
});
