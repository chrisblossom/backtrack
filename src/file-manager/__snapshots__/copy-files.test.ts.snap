// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`copyFiles allowChanges - copies if destFile has not been modified 1`] = `
{
  "<PROJECT_ROOT>/nested/other.js": {
    "dest": "<PROJECT_ROOT>/nested/inside.js",
    "options": {
      "errorOnExist": true,
      "overwrite": true,
      "preserveTimestamps": true,
    },
  },
}
`;

exports[`copyFiles allowChanges - copies if does not exist 1`] = `
{
  "<PROJECT_ROOT>/nested/other.js": {
    "dest": "<PROJECT_ROOT>/nested/missing.js",
    "options": {
      "errorOnExist": true,
      "overwrite": true,
      "preserveTimestamps": true,
    },
  },
}
`;

exports[`copyFiles allowChanges - create file with -latest tag 1`] = `
{
  "<PROJECT_ROOT>/file1.js": {
    "dest": "<PROJECT_ROOT>/file11.js",
    "options": {
      "errorOnExist": true,
      "overwrite": true,
      "preserveTimestamps": true,
    },
  },
  "<PROJECT_ROOT>/nested/other.js": {
    "dest": "<PROJECT_ROOT>/nested/inside.js-latest.js",
    "options": {
      "errorOnExist": true,
      "overwrite": true,
      "preserveTimestamps": true,
    },
  },
}
`;

exports[`copyFiles copies all if all changed 1`] = `
{
  "<PROJECT_ROOT>/file1.js": {
    "dest": "<PROJECT_ROOT>/file11.js",
    "options": {
      "errorOnExist": true,
      "overwrite": true,
      "preserveTimestamps": true,
    },
  },
  "<PROJECT_ROOT>/nested/inside.js": {
    "dest": "<PROJECT_ROOT>/nested/inside1.js",
    "options": {
      "errorOnExist": true,
      "overwrite": true,
      "preserveTimestamps": true,
    },
  },
}
`;

exports[`copyFiles copies only changed files 1`] = `
{
  "<PROJECT_ROOT>/file1.js": {
    "dest": "<PROJECT_ROOT>/file11.js",
    "options": {
      "errorOnExist": true,
      "overwrite": true,
      "preserveTimestamps": true,
    },
  },
}
`;

exports[`copyFiles ignoreUpdates - copies if no previous stats for file 1`] = `
{
  "<PROJECT_ROOT>/nested/other.js": {
    "dest": "<PROJECT_ROOT>/nested/missing.js",
    "options": {
      "errorOnExist": true,
      "overwrite": true,
      "preserveTimestamps": true,
    },
  },
}
`;

exports[`copyFiles ignoreUpdates - does not copy if has previous stats 1`] = `{}`;
